source/util/move_to_front.h

```diff
@@ -0,0 +1,318 @@
+    auto it = id_to_node_.find(id);
+    assert(it != id_to_node_.end());
+    RemoveNode(it->second);
+    // The iterator should still be valid, even if RemoveNode has modified
+    // id_to_node_. But just in case erase by id, not by iterator.
+    id_to_node_.erase(id);
+
+#ifndef NDEBUG
+    deprecated_ids_.insert(id);
+#endif
+  }
+
+  // Returns the number of elements in the move-to-front sequence.
+  size_t GetSize() const {
+    return SizeOf(root_);
+  }
+
+  // The methods below are for testing only.
+
+  // Inserts the value in the internal tree data structure. For testing only.
+  void TestInsert(uint32_t val) {
+    InsertNode(CreateNode(val, val));
+  }
+
+  // Removes the value from the internal tree data structure. For testing only.
+  void TestRemove(uint32_t val) {
+    const auto it = id_to_node_.find(val);
+    assert (it != id_to_node_.end());
+    RemoveNode(it->second);
+  }
+
+  // Prints the internal tree data structure to |out|. For testing only.
+  void PrintTree(std::ostream& out, bool print_timestamp = false) const {
+    if (root_)
+      PrintTreeInternal(out, root_, 1, print_timestamp);
+  }
+
+ private:
+  // Internal tree data structure node.
+  struct Node {
+    // Fields are in order from biggest bit width to smallest.
+    // SPIR-V id.
+    uint32_t id = 0;
+    // Timestamp from a logical clock which updates every time the element is
+    // accessed.
+    uint32_t timestamp = 0;
+    // The size of the node's subtee, including the node.
+    // SizeOf(LeftOf(node)) + SizeOf(RightOf(node)) + 1.
+    uint16_t size = 0;
+    // Handles to connected nodes.
+    uint16_t left = 0;
+    uint16_t right = 0;
+    uint16_t parent = 0;
+    // Distance to the farthest leaf.
+    // Leaves have height 0, real nodes at least 1.
+    uint8_t height = 0;
+  };
+
+  // Creates node and sets correct values. Nodes should be created only through
+  // this function.
+  uint16_t CreateNode(uint32_t timestamp, uint32_t id) {
+    nodes_.emplace_back(Node());
+    Node& node = nodes_.back();
+    node.timestamp = timestamp;
+    node.id = id;
+    node.size = 1;
+    node.height = 1;
+    uint16_t handle = static_cast<uint16_t>(nodes_.size() - 1);
+    id_to_node_.emplace(id, handle);
+    return handle;
+  }
+
+  // Internal tree data structure uses handles instead of pointers. Leaves and
+  // root parent reference a singleton under handle 0. Although dereferencing
+  // a null pointer is not possible, inappropriate access to handle 0 would
+  // cause an assertion. Handles are not garbage collected if id is deprecated.
+  // But handles are recycled when a node is repositioned.
+
+  // Node accessor methods. Naming is designed to be similar to natural
+  // language as these functions tend to be used in sequences, for example:
+  // ParentOf(LeftestDescendentOf(RightOf(node)))
+
+  // Returns immutable id of the node referenced by |handle|.
+  inline uint32_t IdOf(uint16_t handle) const {
```

Comment:
C++11 7.1.2:
para 2:
A function declaration (8.3.5, 9.3, 11.3) with an inline specifier declares an inline function. The inline
specifier indicates to the implementation that inline substitution of the function body at the point of call
is to be preferred to the usual function call mechanism. An implementation is not required to perform this
inline substitution at the point of call; however, even if this inline substitution is omitted, the other rules
for inline functions defined by 7.1.2 shall still be respected.

para 3: A function defined within a class definition is an inline function. 

----
I see I read a little too much into this.   A function defined within a class definition is an inline function, but only has the inline *specifier* if you add it.  And para2 says it's the inline *specifier* that indicates prefer-to-inline.

That being said, I generally shy away from directing the compiler to do optimization stuff that I haven't really proven with tuning data.   Usually, programmers are worse at blind optimization than they think.  For example, there could be a lot of code hidden behind the std::vector::at method (what with bounds checking, exception throwing...).  And when you nest these accessors all together it's not 100% clear what the right thing to do is.  So I personally would leave the inline off and let the compiler make an educated guess based on code size. 

You can keep this if you like, but I personally would remove it.  It does have a read-ability cost for having the extra word.